File: frontend/js/views/two_asset_portfolio.js
// Two-asset portfolio — Markowitz risk/return with correlation, showing the
// diversification benefit, via /calc/two-asset-portfolio. Live as you type.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';

const pct = (n) => Number(n).toLocaleString(undefined, { maximumFractionDigits: 3 }) + '%';
const num = (n) => (n == null ? '—' : Number(n).toLocaleString(undefined, { maximumFractionDigits: 3 }));

export async function renderTwoAssetPortfolio(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.twoasset.h1.title">// TWO-ASSET PORTFOLIO</span></h1>
        <p class="muted small" data-i18n="view.twoasset.hint.intro">
            Risk and return of a two-asset mix. When the assets aren't perfectly correlated, the
            portfolio's volatility falls below the weighted average of the two — the diversification
            benefit. At correlation 1 there's no benefit; at −1 the volatility is minimized. Updates
            as you type.
        </p>
        <div class="lpv-split">
        <div class="chart-panel">
            <h2 data-i18n="view.twoasset.h2.inputs">The mix</h2>
            <form id="twoasset-form" class="inline-form">
                <label><span data-i18n="view.twoasset.label.weighta">Weight of A (%)</span>
                    <input type="number" step="0.1" min="0" max="100" name="weight_a_pct" value="60" required></label>
                <label><span data-i18n="view.twoasset.label.returna">Return A (%)</span>
                    <input type="number" step="0.01" name="return_a_pct" value="8" required></label>
                <label><span data-i18n="view.twoasset.label.returnb">Return B (%)</span>
                    <input type="number" step="0.01" name="return_b_pct" value="4" required></label>
                <label><span data-i18n="view.twoasset.label.vola">Volatility A (%)</span>
                    <input type="number" step="0.01" min="0" name="volatility_a_pct" value="20" required></label>
                <label><span data-i18n="view.twoasset.label.volb">Volatility B (%)</span>
                    <input type="number" step="0.01" min="0" name="volatility_b_pct" value="10" required></label>
                <label><span data-i18n="view.twoasset.label.corr">Correlation (−1 to 1)</span>
                    <input type="number" step="0.01" min="-1" max="1" name="correlation" value="0.2" required></label>
                <label><span data-i18n="view.twoasset.label.rf">Risk-free rate (%)</span>
                    <input type="number" step="0.01" name="risk_free_pct" value="2"></label>
            </form>
        </div>
        <div id="twoasset-result" class="lpv-preview"></div>
        </div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#twoasset-form');
    const generate = async () => {
        const fd = new FormData(form);
        const body = {
            weight_a_pct: Number(fd.get('weight_a_pct')) || 0,
            return_a_pct: Number(fd.get('return_a_pct')) || 0,
            return_b_pct: Number(fd.get('return_b_pct')) || 0,
            volatility_a_pct: Number(fd.get('volatility_a_pct')) || 0,
            volatility_b_pct: Number(fd.get('volatility_b_pct')) || 0,
            correlation: Number(fd.get('correlation')) || 0,
            risk_free_pct: Number(fd.get('risk_free_pct')) || 0,
        };
        try {
            const r = await api.calcTwoAssetPortfolio(body);
            if (!viewIsCurrent(tok)) return;
            renderResult(mount, r);
        } catch (err) {
            showToast(err.message || t('view.twoasset.toast.error'), { level: 'error' });
        }
    };
    form.addEventListener('input', debounce(generate, 250));
    form.addEventListener('submit', (e) => { e.preventDefault(); generate(); });
    generate();
}

function renderResult(mount, r) {
    const el = mount.querySelector('#twoasset-result');
    el.innerHTML = `
        <div class="chart-panel">
            <h2 data-i18n="view.twoasset.h2.result">The portfolio</h2>
            <div class="cards">
                <div class="card pos"><div class="label" data-i18n="view.twoasset.card.return">Return</div>
                    <div class="value pos">${pct(r.portfolio_return_pct)}</div></div>
                <div class="card"><div class="label" data-i18n="view.twoasset.card.vol">Volatility</div>
                    <div class="value">${pct(r.portfolio_volatility_pct)}</div></div>
                <div class="card pos"><div class="label" data-i18n="view.twoasset.card.benefit">Diversification benefit</div>
                    <div class="value pos">${pct(r.diversification_benefit_pct)}</div></div>
            </div>
            <table class="data-table">
                <tbody>
                    <tr><td data-i18n="view.twoasset.row.weights">Weights (A / B)</td><td>${pct(r.weight_a_pct)} / ${pct(r.weight_b_pct)}</td></tr>
                    <tr><td data-i18n="view.twoasset.row.return">Portfolio return</td><td>${pct(r.portfolio_return_pct)}</td></tr>
                    <tr><td data-i18n="view.twoasset.row.vol">Portfolio volatility</td><td>${pct(r.portfolio_volatility_pct)}</td></tr>
                    <tr><td data-i18n="view.twoasset.row.weightedvol">Weighted-avg volatility</td><td>${pct(r.weighted_avg_volatility_pct)}</td></tr>
                    <tr><td data-i18n="view.twoasset.row.sharpe">Sharpe ratio</td><td>${num(r.sharpe_ratio)}</td></tr>
                    <tr class="emph"><td data-i18n="view.twoasset.row.benefit">Diversification benefit</td><td>${pct(r.diversification_benefit_pct)}</td></tr>
                </tbody>
            </table>
        </div>
    `;
    applyUiI18n(el);
}

File: crates/traderview-core/src/two_asset_portfolio.rs
//! Two-asset portfolio risk and return — the Markowitz closed form, showing how
//! correlation below 1 lowers portfolio volatility below the weighted average
//! (the diversification benefit).
//!
//! ```text
//! return = wa·ra + wb·rb
//! var    = wa²σa² + wb²σb² + 2·wa·wb·σa·σb·ρ
//! vol    = √var
//! diversification benefit = (wa·σa + wb·σb) − vol
//! ```
//!
//! At ρ = 1 the portfolio vol equals the weighted average (no benefit); below
//! that it falls, reaching |wa·σa − wb·σb| at ρ = −1.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct TwoAssetInput {
    /// Weight of asset A, percent (B takes the remainder).
    pub weight_a_pct: f64,
    pub return_a_pct: f64,
    pub return_b_pct: f64,
    pub volatility_a_pct: f64,
    pub volatility_b_pct: f64,
    /// Correlation between the assets, −1..1.
    pub correlation: f64,
    /// Risk-free rate for the Sharpe ratio, percent.
    #[serde(default)]
    pub risk_free_pct: f64,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct TwoAssetResult {
    pub weight_a_pct: f64,
    pub weight_b_pct: f64,
    /// Expected portfolio return, percent.
    pub portfolio_return_pct: f64,
    /// Portfolio volatility (standard deviation), percent.
    pub portfolio_volatility_pct: f64,
    /// Weight-averaged volatility — the no-diversification benchmark.
    pub weighted_avg_volatility_pct: f64,
    /// weighted_avg − portfolio vol; the volatility shaved off by diversifying.
    pub diversification_benefit_pct: f64,
    /// (return − risk-free) / volatility; `None` if volatility is 0.
    pub sharpe_ratio: Option<f64>,
}

pub fn analyze(input: &TwoAssetInput) -> TwoAssetResult {
    let wa = input.weight_a_pct / 100.0;
    let wb = 1.0 - wa;
    let sa = input.volatility_a_pct / 100.0;
    let sb = input.volatility_b_pct / 100.0;
    let rho = input.correlation.clamp(-1.0, 1.0);

    let ret = wa * input.return_a_pct + wb * input.return_b_pct;

    let var = wa * wa * sa * sa + wb * wb * sb * sb + 2.0 * wa * wb * sa * sb * rho;
    let vol = var.max(0.0).sqrt() * 100.0;
    let weighted_avg = (wa * sa + wb * sb) * 100.0;

    let sharpe = if vol > 0.0 {
        Some((ret - input.risk_free_pct) / vol)
    } else {
        None
    };

    TwoAssetResult {
        weight_a_pct: input.weight_a_pct,
        weight_b_pct: wb * 100.0,
        portfolio_return_pct: ret,
        portfolio_volatility_pct: vol,
        weighted_avg_volatility_pct: weighted_avg,
        diversification_benefit_pct: weighted_avg - vol,
        sharpe_ratio: sharpe,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn close(a: f64, b: f64) -> bool {
        (a - b).abs() < 1e-4
    }

    fn run(wa: f64, rho: f64) -> TwoAssetResult {
        analyze(&TwoAssetInput {
            weight_a_pct: wa,
            return_a_pct: 8.0,
            return_b_pct: 4.0,
            volatility_a_pct: 20.0,
            volatility_b_pct: 10.0,
            correlation: rho,
            risk_free_pct: 0.0,
        })
    }

    #[test]
    fn portfolio_return() {
        // 0.6×8 + 0.4×4 = 6.4%.
        assert!(close(run(60.0, 0.2).portfolio_return_pct, 6.4));
    }

    #[test]
    fn portfolio_volatility() {
        // var = 0.0144 + 0.0016 + 0.00192 = 0.01792 → vol 13.3866%.
        assert!(close(run(60.0, 0.2).portfolio_volatility_pct, 13.3866));
    }

    #[test]
    fn weighted_average_volatility() {
        assert!(close(run(60.0, 0.2).weighted_avg_volatility_pct, 16.0));
    }

    #[test]
    fn diversification_benefit() {
        // 16 − 13.3866 = 2.6134.
        assert!(close(run(60.0, 0.2).diversification_benefit_pct, 2.6134));
    }

    #[test]
    fn perfect_correlation_no_benefit() {
        let r = run(60.0, 1.0);
        assert!(close(r.portfolio_volatility_pct, 16.0));
        assert!(close(r.diversification_benefit_pct, 0.0));
    }

    #[test]
    fn negative_correlation_minimizes_vol() {
        // ρ = −1 → |0.6×20 − 0.4×10| = |12 − 4| = 8%.
        assert!(close(run(60.0, -1.0).portfolio_volatility_pct, 8.0));
    }

    #[test]
    fn sharpe_ratio() {
        let r = analyze(&TwoAssetInput {
            weight_a_pct: 60.0,
            return_a_pct: 8.0,
            return_b_pct: 4.0,
            volatility_a_pct: 20.0,
            volatility_b_pct: 10.0,
            correlation: 0.2,
            risk_free_pct: 2.0,
        });
        // (6.4 − 2) / 13.3866 = 0.328686.
        assert!(close(r.sharpe_ratio.unwrap(), 0.328686));
    }

    #[test]
    fn single_asset_collapses_to_asset_a() {
        let r = run(100.0, 0.2);
        assert!(close(r.portfolio_return_pct, 8.0));
        assert!(close(r.portfolio_volatility_pct, 20.0));
        assert!(close(r.diversification_benefit_pct, 0.0));
    }
}
